src/evict/evict_lru.c

```diff
@@ -1385,9 +1385,15 @@ __evict_get_ref(
 	if (is_server && candidates > 1)
 		candidates /= 2;
 
-	/* Get the next page queued for eviction. */
+	/*
+	 * Get the next page queued for eviction: check both the WT_REF and the
+	 * page's flag: they're not set atomically, we can race with the server
+	 * setting them. Check the page's flag last, it's the element set last
+	 * and atomically.
+	 */
 	while ((evict = cache->evict_current) != NULL &&
-	    evict < cache->evict_queue + candidates && evict->ref != NULL) {
+	    evict < cache->evict_queue + candidates && evict->ref != NULL &&
+	    F_ISSET_ATOMIC(evict->ref->page, WT_PAGE_EVICT_LRU)) {
```

Comment:
This makes me a bit nervous. Will a page stay on the eviction queue forever if it is added without the ``WT_PAGE_EVICT_LRU`` flag set? i.e: I'd rather be able to assert that the flag is always set, rather than use it as a condition to break out of here.